sqlmesh/migrations/v0028_generate_schema_types_using_dialect.py

```diff
@@ -0,0 +1,68 @@
+"""Generate mapping schema data types using the corresponding model's dialect."""
+import json
+
+import pandas as pd
+from sqlglot import exp, parse_one
+
+from sqlmesh.utils.migration import index_text_type
+
+
+def migrate(state_sync):  # type: ignore
+    engine_adapter = state_sync.engine_adapter
+    schema = state_sync.schema
+    snapshots_table = "_snapshots"
+    if schema:
+        snapshots_table = f"{schema}.{snapshots_table}"
+
+    def _convert_schema_types(schema):
+        if not schema:
+            return schema
+
+        for k, v in schema.items():
+            if isinstance(v, dict):
+                _convert_schema_types(v)
```

Comment:
why is this nested?

Reply:
as in "why it's defined as a closure" or the recursion part?

Reply:
recursion

Reply:
Since the schemas are nested and we need to update their leaves, this seemed like a simple way to do it, considering that the nesting levels shouldn't be deeper than 2-3. Do you think this is problematic?

Reply:
FYI the [maximum nesting level in BigQuery is 15](https://cloud.google.com/bigquery/docs/nested-repeated) - I can't speak for other engines however.

Reply:
I think that nesting takes into account records / structs, right? We don't nest these IIRC, they're simply represented as strings like `STRUCT<foo STRUCT<... >>`, so it should be fine.